hwilib/devices/bitbox02.py

```diff
@@ -0,0 +1,681 @@
+                prevtx = psbt_in.non_witness_utxo
+            elif psbt_in.witness_utxo:
+                utxo = psbt_in.witness_utxo
+            if utxo is None:
+                raise BadArgumentError("No utxo found for input {}".format(input_index))
+            if prevtx is None:
+                raise BadArgumentError(
+                    "Previous transaction missing for input {}".format(input_index)
+                )
+
+            found_pubkey, keypath = find_our_key(psbt_in.hd_keypaths)
+            if not found_pubkey:
+                raise BadArgumentError("No key found for input {}".format(input_index))
+            assert keypath is not None
+            found_pubkeys.append(found_pubkey)
+
+            # TOOD: validate keypath
+
+            if bip44_account is None:
+                bip44_account = keypath[2]
+            elif bip44_account != keypath[2]:
+                raise BadArgumentError(
+                    "The bip44 account index must be the same for all inputs and changes"
+                )
+
+            simple_type = get_simple_type(utxo, psbt_in.redeem_script)
+
+            script_config_index_map = {
+                bitbox02.btc.BTCScriptConfig.P2WPKH: 0,
+                bitbox02.btc.BTCScriptConfig.P2WPKH_P2SH: 1,
+            }
+
+            inputs.append(
+                {
+                    "prev_out_hash": ser_uint256(tx_in.prevout.hash),
+                    "prev_out_index": tx_in.prevout.n,
+                    "prev_out_value": utxo.nValue,
+                    "sequence": tx_in.nSequence,
+                    "keypath": keypath,
+                    "script_config_index": script_config_index_map[simple_type],
+                    "prev_tx": {
+                        "version": prevtx.nVersion,
+                        "locktime": prevtx.nLockTime,
+                        "inputs": [
+                            {
+                                "prev_out_hash": ser_uint256(prev_in.prevout.hash),
+                                "prev_out_index": prev_in.prevout.n,
+                                "signature_script": prev_in.scriptSig,
+                                "sequence": prev_in.nSequence,
+                            }
+                            for prev_in in prevtx.vin
+                        ],
+                        "outputs": [
+                            {
+                                "value": prev_out.nValue,
+                                "pubkey_script": prev_out.scriptPubKey,
+                            }
+                            for prev_out in prevtx.vout
+                        ],
+                    },
+                }
+            )
+
+        outputs: List[bitbox02.BTCOutputType] = []
+        for output_index, (psbt_out, tx_out) in builtins.enumerate(
+            zip(psbt.outputs, psbt.tx.vout)
+        ):
+            _, keypath = find_our_key(psbt_out.hd_keypaths)
+            is_change = keypath and keypath[-2] == 1
```

Comment:
Sending the output as `bitbox02.BTCOutputInternal` tells the device that it is a change output. 

The BitBox02 enforces that change outputs have the `change` part of the keypath be `1`, hence this check. 

By adding this check, the output is shown to the user. Removing this check would mean the whole transaction will fail.

I guess if there was a tx that sent the change on a strange keypath, we should let it fail instead of showing an address to the user that they cannot effectively verify. So I will remove this additional check.

Reply:
I took a look, I have to keep the check. The reason is that Electrum (and others?) seem to include the keypath even for non-change outputs if you send to yourself, so the presence of a keypath in the output is not sufficient to determine if it actually is a change (automatically added), or a deliberate recipient address.

The other HW wallets seem to treat it as change and not display it on their screens, which is totally fine, but the BitBox02 shows all outputs manually added by a user for verification, even if it is a send-to-self.